### PR TITLE
Sphere-to-3d

### PR DESCRIPTION
**Pull request by selipot**
Adding two functions that invert each other to the sphere module: sphere_to_threedim and threedim_to_sphere

These functions will be used by spheretrans which is the wavelet transform on a sphere which is eventually needed for the upcoming extraction of inertial oscillations from geographical positions.

Open to name change etc.

### CODE REVIEW

**Comment by philippemiron** [comment]
If we are discussing double/single precision, maybe this could be done on an ellipsoid, the equations do not look much more complex, see [here](https://gssc.esa.int/navipedia/index.php/Ellipsoidal_and_Cartesian_Coordinates_Conversion). I don't know how much difference that would make.

image

**Comment by selipot** [comment]
That could be a good thing to implement. However, in equation (1) on that page, if the excentricity `e = 0` then the coefficient in front is `2h` which is not what our current equations predict. What am I missing?

**Comment by selipot** [comment]
I think I was missing that `h` is maybe not the Earth radius but the third ellipsoidal coordinate.

**Comment by philippemiron** [comment]
Yes `h` is the height above/below the ellipsoid.

**Comment by selipot** [comment]
I think @philippemiron 's suggestion to implement the ellipsoidal case is a good one but maybe one for another PR/set of functions.

**Comment by milancurcic** [comment]
Ellipsoid sounds great.

File: clouddrift/sphere.py [hunk 1]
```diff
@@ -435,3 +435,133 @@ def sphere_to_plane(
     y[..., :] = np.cumsum(dy, axis=-1)
 
     return x, y
+
+
+def earthsphere_to_cartesian(
```
Comment (milancurcic) [inline clouddrift/sphere.py:440]: Why not `sphere_to_cartesian`? Unless the ellipsoid stuff will be Earth-specific. But it seems to me that these functions will be useful on other planets too, e.g. Mars.

File: clouddrift/sphere.py [hunk 2]
```diff
@@ -435,3 +435,133 @@ def sphere_to_plane(
...
+    lon: np.ndarray,
+    lat: np.ndarray,
+    earth_radius: Optional[float] = EARTH_RADIUS_METERS,
```
Comment (milancurcic) [inline clouddrift/sphere.py:443]: Same here, can be just `radius` that defaults to `EARTH_RADIUS_METERS`.

File: clouddrift/sphere.py [hunk 3]
```diff
@@ -435,3 +435,133 @@ def sphere_to_plane(
...
+        If ``lon`` and ``lat`` are not NumPy arrays.
+
+    See Also
+    --------
+    :func:`cartesian_to_earthsphere`
+    """
+
+    x = earth_radius * np.cos(lat * np.pi / 180) * np.cos(lon * np.pi / 180)
+    y = earth_radius * np.cos(lat * np.pi / 180) * np.sin(lon * np.pi / 180)
+    z = earth_radius * np.sin(lat * np.pi / 180)
```
Comment (milancurcic) [inline clouddrift/sphere.py:497]:
```suggestion
    lonr, latr = np.deg2rad(lon), np.deg2rad(lat)

    x = earth_radius * np.cos(latr) * np.cos(lonr)
    y = earth_radius * np.cos(latr) * np.sin(lonr)
    z = earth_radius * np.sin(latr)
```

File: clouddrift/sphere.py [hunk 4]
```diff
@@ -435,3 +435,133 @@ def sphere_to_plane(
...
+
+    x = np.nan_to_num(x, posinf=np.nan, neginf=np.nan)
+    y = np.nan_to_num(y, posinf=np.nan, neginf=np.nan)
+    z = np.nan_to_num(z, posinf=np.nan, neginf=np.nan)
```
Comment (milancurcic) [inline clouddrift/sphere.py:501]: I don't think this is necessary. For what values of lon, lat do we expect +/- inf in x, y, z? It seems to me that the result values are always bounded by the radius.

File: clouddrift/sphere.py [hunk 5]
```diff
@@ -435,3 +435,133 @@ def sphere_to_plane(
...
+    :func:`earthsphere_to_cartesian`
+    """
+
+    R = np.sqrt(np.abs(x) ** 2 + np.abs(y) ** 2 + np.abs(z) ** 2)
+    x /= R
+    y /= R
+    z /= R
+
+    lon = recast_lon180(np.imag(np.log((x + 1j * y)))) * 180 / np.pi
+    lat = np.arcsin(z) * 180 / np.pi
```
Comment (milancurcic) [inline clouddrift/sphere.py:565]: As above

```suggestion
    lon = np.rad2deg(recast_lon180(np.imag(np.log((x + 1j * y)))))
    lat = np.rad2deg(np.arcsin(z))
```

File: clouddrift/sphere.py [hunk 6]
```diff
@@ -435,3 +435,133 @@ def sphere_to_plane(
...
+    ------
+    AttributeError
+        If ``x``, ``y``, and ``z`` are not NumPy arrays.
+
+    See Also
+    --------
+    :func:`earthsphere_to_cartesian`
+    """
+
+    R = np.sqrt(np.abs(x) ** 2 + np.abs(y) ** 2 + np.abs(z) ** 2)
```
Comment (milancurcic) [inline clouddrift/sphere.py:559]: Are `abs()` needed? It seems to me they make a difference only when inputs are complex; is that expected?

File: clouddrift/sphere.py (hunk 4 above)
Comment (milancurcic) [reply clouddrift/sphere.py:501]: If we remove this, the function is compatible with `xr.DataArray` as well.

**Review by milancurcic — APPROVED** [review]

File: clouddrift/sphere.py (hunk 1 above)
Comment (selipot) [reply clouddrift/sphere.py:440]: ok!

File: clouddrift/sphere.py (hunk 2 above)
Comment (selipot) [reply clouddrift/sphere.py:443]: ok!

File: clouddrift/sphere.py (hunk 3 above)
Comment (philippemiron) [inline clouddrift/sphere.py:497]: I think using numpy `np.radians` makes things cleaner.

```suggestion
    x = earth_radius * np.cos(np.radians(lat)) * np.cos(np.radians(lat))
    y = earth_radius * np.cos(np.radians(lat)) * np.sin(np.radians(lon))
    z = earth_radius * np.sin(np.radians(lat))
```

File: clouddrift/sphere.py [hunk 7]
```diff
@@ -435,3 +435,133 @@ def sphere_to_plane(
...
+    x /= R
+    y /= R
+    z /= R
+
+    lon = recast_lon180(np.imag(np.log((x + 1j * y)))) * 180 / np.pi
```
Comment (philippemiron) [inline clouddrift/sphere.py:564]:
```suggestion
    lon = np.radians(recast_lon180(np.imag(np.log((x + 1j * y)))))
```

Comment (philippemiron) [inline clouddrift/sphere.py:564]:
```suggestion
    lon = np.degrees(recast_lon180(np.imag(np.log((x + 1j * y)))))
```

File: clouddrift/sphere.py (hunk 5 above)
Comment (philippemiron) [inline clouddrift/sphere.py:565]:
```suggestion
    lat = np.degrees(np.arcsin(z))
```

**Review by philippemiron — APPROVED** [review]

File: clouddrift/sphere.py (hunk 3 above)
Comment (selipot) [reply clouddrift/sphere.py:497]: ok!

**Comment by philippemiron** [comment]
I guess we suggested the same thing ... Ps: Personally, I prefer `np.radians` and `np.degrees` than the `np.deg2rad` `np.rad2deg` weird name style.

File: clouddrift/sphere.py (hunk 7 above)
Comment (selipot) [reply clouddrift/sphere.py:564]: You need to convert to degree before recasting?

File: clouddrift/sphere.py (hunk 5 above)
Comment (selipot) [reply clouddrift/sphere.py:565]: ok

**Comment by milancurcic** [comment]
@philippemiron Cool! I didn't know about `np.radians` and `np.degrees`.

File: clouddrift/sphere.py (hunk 5 above)
Comment (selipot) [reply clouddrift/sphere.py]: Let's use `np.deg2rad` and `np.rad2deg` as it is already used in other functions.

File: clouddrift/sphere.py [hunk 8]
```diff
@@ -195,20 +195,23 @@ def recast_lon(lon: np.ndarray, lon0: Optional[float] = -180) -> np.ndarray:
     Returns
     -------
     np.ndarray or float
-        Converted longitudes in the range `[lon0, lon0+360]`
+        Converted longitudes in the range `[lon0, lon0+360[`
```
Comment (milancurcic) [inline clouddrift/sphere.py:198]: Should it not be `[lon0, lon0+360)` ?

Comment (selipot) [reply clouddrift/sphere.py:198]: Both seem accepted https://en.wikipedia.org/wiki/ISO_31-11

Comment (milancurcic) [reply clouddrift/sphere.py:198]: Interesting, I didn't know.

**Pull request merged**